### ISSUE TITLE
[Cache] Directory does not exist

### ISSUE THREAD

**Issue by Oliboy50** [symfony/symfony#20517]
## Details
I'm facing the following error message

```
Cache directory does not exist (/path/to/my/symfony_project/var/cache/prod/pools/ngRSYeFcCP)
```

on 2 different projects, on production environment, running on multiple servers, with a high traffic.

~~One project is running Symfony 3.1.5, the other one is running Symfony 3.1.4~~
Both projects are running Symfony 3.1.7

This error happens **ONLY** during new deployments (new releases are reachable only after `cache:warmup`).

The error is thrown when someone calls one of the project routes (the exception is not thrown on `cache:warmup`)... so it returns a `500` to the user, which is not so nice, even if it occurs only once :)

## Questions
- Is this possible that I'm facing a race condition between [these lines](https://github.com/symfony/symfony/blob/v3.1.6/src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php#L36-L40)?
- ~~Does this change in the condition: [`... ?: (file_exists($dir) ? $dir : false))`](https://github.com/symfony/symfony/blob/3.1/src/Symfony/Component/Cache/Adapter/FilesystemAdapter.php#L39) will fix this issue? And if so, will this be available in 3.1.7?~~

Thank you!

**Comment by patrick-mcdougle** [comment]
Look into the real path cache. I know that has bit us in deploying our PHP websites.

**Comment by nicolas-grekas** [comment]
So, does 3.1.7 fix the issue for you?
Otherwise, can you try by replacing `'/.'` by `'/'`? (removing the dot) a few lines above?

**Comment by Oliboy50** [comment]
we'll try this within the next few days, and i'll give you feedbacks here

**Comment by Oliboy50** [comment]
@nicolas-grekas `3.1.7` does not fix this issues

But I have more information about this:

- This only occurs during deployments on multiple servers 
- This can occur multiple times during a single deployment (the most we've seen is 3 errors during a deployment on 15 servers... each error thrown by a different server... errors occured in a time period of less than 100ms).
- This is probably due to the way we invalidate opcache during deployments.

I'm not sure we'll have time to fix this, because it's considered as a minor bug on our side, but if we do, I'll give you feedbacks here again.

**Comment by nicolas-grekas** [comment]
You're warming up the cache while the server is running? Are your clearing the cache beforehand? How? Can you describe your deployment process so that we can figure out how the race condition is met a bit?

**Comment by teohhanhui** [comment]
I think I'm running into a similar issue:

> Unable to create the Doctrine Proxy directory

It also only happens (sometimes) on the first requests after cache warmup. Subsequent requests after a certain time window work just fine, being able to create the proxy directory as expected.

**Comment by nicolas-grekas** [comment]
@teohhanhui @Oliboy50 could you please tell us if you're warming up the cache while the server is running? Are your clearing the cache beforehand? How? Can you describe your deployment process so that we can figure out how the race condition is met a bit?

**Comment by Oliboy50** [comment]
@nicolas-grekas we use Capistrano as our deployment system, therefore the whole project is built (along with the cache) in a separate directory then, if everything worked fine, the symlink is updated to point to the new project sources

**Comment by teohhanhui** [comment]
It was `rm -r var/cache/*` followed by `composer install ...`. (Basically
Symfony's composer post-install script defaults to `--no-warmup` when doing
the `cache:clear`.) It's in a Docker volume. Symlink is not used.

So, sometimes (very rarely?) the error I described above happens on the
first page load, but I'm not sure how to reproduce.

**Comment by luislaborda** [comment]
I'm having the exact same problem. When running the command php bin/console cache:clear --env=prod I get the following error. Have you guys found the solution? I'm stock and I cannot deploy, site is down.

[Symfony\Component\Cache\Exception\InvalidArgumentException]                           
 Cache directory does not exist (/u02/websys/symfony/cache/prcr/prod/pools/CktlTHB2Cy)

**Comment by nicolas-grekas** [comment]
@luislaborda does it happen consistently to you? Because the previous report was saying this happens sometimes. Would you be able to debug the situation? We can't figure out what can be wrong currently.

**Comment by ramrami** [comment]
We're running into a similar issue, with symfony v3.2.6.
Sometimes (very rarely),  while running the tests we get `InvalidArgumentException: Cache directory does not exist` when executing this line `self::bootKernel(['debug' => false]);`

We're running the tests in parallel so maybe it's a race condition as @Oliboy50 said.

**Comment by nicolas-grekas** [comment]
@ramrami runnng tests in parallel creates an obvious race condition if could clean the test directory while another concurrent test creates it. So may not be the same issue.
Looking at the component, it does never remove directories, so might no be related to the component itself, but to some wiring code that removes it.

**Comment by nicolas-grekas** [comment]
Should be fixed by #22374
